src/script/lua_api/l_object.cpp

```diff
@@ -1410,6 +1410,140 @@ int ObjectRef::l_get_player_control_bits(lua_State *L)
 	return 1;
 }
 
+// Reads in a physics modifier from a table, defaulting to existing values.
+// Returns success.
+static bool read_physics_modifier(lua_State *L, int index,
+		PlayerSAO::PhysicsModifier &modifier)
+{
+	if (!lua_istable(L, index))
+		return false;
+
+	auto type = getstringfield_default(L, index, "type", "multiply");
+	if (type == "add") {
+		modifier.is_add = true;
+	} else if (type != "multiply") {
+		warningstream << "Unknown modifier type " << type << ", defaulting to multiply." << std::endl;
+		warningstream << script_get_backtrace(L);
+	}
+
+	int def =  modifier.is_add ? 0 : 1;
+
+	modifier.speed = getfloatfield_default(L, index, "speed", def);
+	modifier.jump = getfloatfield_default(L, index, "jump", def);
+	modifier.gravity = getfloatfield_default(L, index, "gravity", def);
+
+	return true;
+}
+
+// get_effective_physics(self)
+int ObjectRef::l_get_effective_physics(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	PlayerSAO *playersao = (PlayerSAO *)getobject(ref);
+	if (playersao == nullptr)
+		return 0;
+
+	const auto &physics = playersao->getEffectivePhysics();
+
+	lua_newtable(L);
+	lua_pushnumber(L, physics.speed);
+	lua_setfield(L, -2, "speed");
+	lua_pushnumber(L, physics.jump);
+	lua_setfield(L, -2, "jump");
+	lua_pushnumber(L, physics.gravity);
+	lua_setfield(L, -2, "gravity");
+	return 1;
+}
+
+// set_physics_modifier(self, key, modifier_table)
+int ObjectRef::l_set_physics_modifier(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	PlayerSAO *playersao = (PlayerSAO *) getobject(ref);
+	if (playersao == nullptr)
+		return 0;
+
+	const std::string key = readParam<std::string>(L, 2);
+
+	PlayerSAO::PhysicsModifier modifier;
+	if (!read_physics_modifier(L, 3, modifier))
+		throw LuaError("set_physics_modifier expects (string, table)");
+
+	playersao->setPhysicsModifier(key, modifier);
+
+	if (playersao->m_physics_override_set) {
+		warningstream << "set_physics_modifier will have no effect because ";
+		warningstream << "set_physics_override was previously called" << std::endl;
+		warningstream << script_get_backtrace(L);
```

Comment:
at least the backtrace here should be moved to infostream
(the mod calling `set_physics_modifier` did nothing wrong)